tools/c7n_mailer/c7n_mailer/azure_mailer/sendgrid_delivery.py

```diff
@@ -22,11 +23,12 @@
 
 class SendGridDelivery(object):
 
-    def __init__(self, config, logger):
+    def __init__(self, config, session, logger):
         self.config = config
         self.logger = logger
-        self.sendgrid_client = \
-            sendgrid.SendGridAPIClient(self.config.get('sendgrid_api_key', ''))
+        self.session = session
+        api_key = decrypt(self.config, self.logger, self.session, 'sendgrid_api_key')
```

Comment:
rather than having each component decrypt/encrypt, and having to propagate cloud sessions, I feel like we should have a config initialization that does this once up front. thoughts?

Reply:
I think our long-term plan is to leverage c7n plugin registry for this. Registry will do cache, sessions etc in a proper way. 

I agree that for mailer config initialization is possible, but I don't think it is possible for c7n in general (e.g. some providers will leverage resources in the runtime to extract some value from tags).

Reply:
re: sessions. I think we still need to propagate session to mailer delivery plugins to maintain common interface for delivery plugins.

Reply:
okay, sounds good.